crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -3009,3 +3036,54 @@ pub(crate) fn check_ast(
 
     (diagnostics.into_inner(), semantic_errors.into_inner())
 }
+
+/// An abstraction for mutating a diagnostic.
+///
+/// Callers can build this guard by starting with `Checker::report_diagnostic`.
+///
+/// The primary function of this guard is to add the underlying diagnostic to the `Checker`'s list
+/// of diagnostics on `Drop`, while dereferencing to the underlying diagnostic for mutations like
+/// adding fixes or parent ranges.
+pub(crate) struct DiagnosticGuard<'a, 'b> {
+    /// The parent checker that will receive the diagnostic on `Drop`.
+    checker: &'a Checker<'b>,
+    /// The diagnostic that we want to report.
+    ///
+    /// This is always `Some` until the `Drop` (or `defuse`) call.
+    diagnostic: Option<Diagnostic>,
+}
+
+impl DiagnosticGuard<'_, '_> {
+    /// Consume the underlying `Diagnostic` without emitting it.
+    ///
+    /// In general you should avoid constructing diagnostics that may not be emitted, but this
+    /// method can be used where this is unavoidable.
+    pub(crate) fn defuse(mut self) {
+        self.diagnostic = None;
+    }
+}
+
+impl std::ops::Deref for DiagnosticGuard<'_, '_> {
+    type Target = Diagnostic;
+
+    fn deref(&self) -> &Diagnostic {
+        // OK because `self.diagnostic` is only `None` within `Drop`.
+        self.diagnostic.as_ref().unwrap()
+    }
+}
+
+/// Return a mutable borrow of the diagnostic in this guard.
+impl std::ops::DerefMut for DiagnosticGuard<'_, '_> {
+    fn deref_mut(&mut self) -> &mut Diagnostic {
+        // OK because `self.diagnostic` is only `None` within `Drop`.
+        self.diagnostic.as_mut().unwrap()
+    }
+}
+
+impl Drop for DiagnosticGuard<'_, '_> {
+    fn drop(&mut self) {
+        if let Some(diagnostic) = self.diagnostic.take() {
+            self.checker.diagnostics.borrow_mut().push(diagnostic);
+        }
```

Comment:
we shouldn't emit the diagnostic if the thread is unwinding:

```suggestion
        if std::thread::panicking {
            return;
        }
        
        if let Some(diagnostic) = self.diagnostic.take() {
            self.checker.diagnostics.borrow_mut().push(diagnostic);
        }
```

Reply:
Makes sense, I saw that in ty but thought it might be ty-specific. Do you know if I also need the second one after the `take` call?

https://github.com/astral-sh/ruff/blob/839c637abb8c7f684f7577885a53432bff949169/crates/ty_python_semantic/src/types/context.rs#L503-L516

Reply:
Hmm no, I think that's just wrong.